pkg/controller/machine/controller.go

```diff
@@ -234,25 +289,57 @@ func (r *ReconcileMachine) Reconcile(request reconcile.Request) (reconcile.Resul
 		return reconcile.Result{}, nil
 	}
 
-	exist, err := r.actuator.Exists(ctx, cluster, m)
+	instanceExist, err := r.actuator.Exists(ctx, cluster, m)
 	if err != nil {
 		klog.Errorf("Failed to check if machine %q exists: %v", name, err)
 		return reconcile.Result{}, err
 	}
 
-	if exist {
+	if instanceExist {
 		klog.Infof("Reconciling machine %q triggers idempotent update", name)
 		if err := r.actuator.Update(ctx, cluster, m); err != nil {
 			klog.Errorf(`Error updating machine "%s/%s": %v`, m.Namespace, name, err)
 			return delayIfRequeueAfterError(err)
 		}
+
+		if !machineIsProvisioned(m) {
+			klog.Errorf(`Instance for Machine "%s/%s exists but providerID or addresses has not been given to the machine yet"`, m.Namespace, name)
+			return reconcile.Result{}, err
+		}
+		if machineHasNode(m) {
+			if err := r.setPhase(m, phaseRunning, ""); err != nil {
+				return reconcile.Result{}, err
+			}
+		} else {
+			if err := r.setPhase(m, phaseProvisioned, ""); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
 		return reconcile.Result{}, nil
 	}
 
-	// Machine resource created. Machine does not yet exist.
+	// Instance does not exist but the machine has been given a providerID/address.
+	// This can only be reached if an instance was deleted outside the machine API
+	if machineIsProvisioned(m) {
+		if err := r.setPhase(m, phaseFailed, "Can't find created instance."); err != nil {
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, nil
+	}
+
+	// Machine resource created and instance does not exist yet.
+	if err := r.setPhase(m, phaseProvisioning, ""); err != nil {
+		return reconcile.Result{}, err
+	}
 	klog.Infof("Reconciling machine object %v triggers idempotent create.", m.ObjectMeta.Name)
 	if err := r.actuator.Create(ctx, cluster, m); err != nil {
 		klog.Warningf("Failed to create machine %q: %v", name, err)
+		if isInvalidMachineConfigurationError(err) {
+			if err := r.setPhase(m, phaseFailed, err.Error()); err != nil {
```

Comment:
Isn't the point of setting this to prevent further reconciliation? Where are we doing that?

Reply:
in line 341

Reply:
I think I was misunderstanding the intent here. I  guess a machine is allowed to come out of a `Failed` state *if* an instance has not yet been provisioned. Otherwise it will be forever `Failed`.

Reply:
https://github.com/openshift/cluster-api/pull/75/commits/8f3e6ac87f353cdef46d6926a25ce627d5da9d3c#diff-b96ce5bf8686fb02d855188000ae1cb4R272